File: KevinLeonard/QL/src/main/scala/ast/Expression.scala
package ast

import scala.util.parsing.input.Positional

// Boolean and arithmetic expressions
sealed trait Expression extends Positional

case class Or(l: Expression, r: Expression) extends Expression
case class And(l: Expression, r: Expression) extends Expression
case class Not(e: Expression) extends Expression
case class Equal(l: Expression, r: Expression) extends Expression
case class NotEqual(l: Expression, r: Expression) extends Expression
case class LessThan(l: Expression, r: Expression) extends Expression
case class LessThanEqual(l: Expression, r: Expression) extends Expression
case class GreaterThan(l: Expression, r: Expression) extends Expression
case class GreaterThanEqual(l: Expression, r: Expression) extends Expression
case class Add(l: Expression, r: Expression) extends Expression
case class Sub(l: Expression, r: Expression) extends Expression
case class Mul(l: Expression, r: Expression) extends Expression
case class Div(l: Expression, r: Expression) extends Expression
case class Variable(name: String) extends Expression

sealed trait Literal extends Expression
case class BooleanLiteral(value: Boolean) extends Literal
case class NumberLiteral(value: Int) extends Literal
case class StringLiteral(value: String) extends Literal

File: geert.kai/lib/ql/ast/expression.rb
require_relative "form"

module QL
  module AST
    class Expression < Node
      def accept(visitor)
        visitor.visit(self)
      end
    end

    class Variable < Expression
      attr_reader :name

      def initialize(name)
        @name = name
      end
    end

    class Literal < Expression
      attr_reader :value

      def initialize(value)
        @value = value
      end

      def to_s
        @value.to_s
      end
    end

    class IntegerLiteral < Literal
      def type
        IntegerType.new
      end
    end

    class StringLiteral < Literal
      def type
        StringType.new
      end
    end

    class BooleanLiteral < Literal
      def type
        BooleanType.new
      end
    end

    # Expressions

    class BinaryExpression < Expression
      attr_reader :lhs, :rhs

      def initialize(lhs, rhs)
        @lhs = lhs
        @rhs = rhs
      end
    end

    class BooleanExpression < BinaryExpression # || &&
      def type
        BooleanType.new
      end

      def possible_argument_types
        [BooleanType.new]
      end
    end

    class EqualityExpression < BinaryExpression # ==, !=
      def type
        BooleanType.new
      end

      def possible_argument_types
        [BooleanType.new, StringType.new, IntegerType.new]
      end
    end

    class OrderingExpression < BinaryExpression # <, >, <=, >=
      def type
        BooleanType.new
      end

      def possible_argument_types
        [IntegerType.new]
      end
    end

    class IntegerExpression < BinaryExpression # * + - /
      def type
        IntegerType.new
      end

      def possible_argument_types
        [IntegerType.new]
      end
    end

    class And < BooleanExpression
      def operator
        :&
      end
    end

    class Or < BooleanExpression
      def operator
        :|
      end
    end

    class Equal < EqualityExpression
      def operator
        :==
      end
    end

    class Inequal < EqualityExpression
      def operator
        :!=
      end
    end

    class LessThan < OrderingExpression
      def operator
        :<
      end
    end

    class LessThanOrEqualTo < OrderingExpression
      def operator
        :<=
      end
    end

    class GreaterThan < OrderingExpression
      def operator
        :>
      end
    end

    class GreaterThanOrEqualTo < OrderingExpression
      def operator
        :>=
      end
    end

    class Plus < IntegerExpression
      def operator
        :+
      end
    end

    class Minus < IntegerExpression
      def operator
        :-
      end
    end

    class Multiplication < IntegerExpression
      def operator
        :*
      end
    end

    class Division < IntegerExpression
      def operator
        :/
      end
    end
  end
end